spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/CassandraRow.scala

```diff
@@ -72,12 +72,65 @@ import com.datastax.driver.core.Row
   *   - java.sql.Date
   *   - org.joda.time.DateTime
   */
-final class CassandraRow(val columnNames: IndexedSeq[String], val columnValues: IndexedSeq[AnyRef])
+final class CassandraRow(val metaData:CassandraRowMetaData, val columnValues: IndexedSeq[AnyRef])
   extends ScalaGettableData with Serializable {
 
+  /**
+    * the consturctor is for testing and backward compatibility only.
+    * Use default constructor with shared metadata for memory saving and performance.
+    *
+    * @param columnNames
+    * @param columnValues
+    */
+  def this (columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef]) =
+    this(CassandraRowMetaData.fromColumnNames(columnNames), columnValues)
+
   override def toString = "CassandraRow" + dataAsString
 }
 
+case class CassandraRowMetaData(columnNames: IndexedSeq[String], resultSetColumnNames:Option[IndexedSeq[String]] = None, @transient codecs: Option[IndexedSeq[TypeCodec[AnyRef]]] = None) {
```

Comment:
codecs are marked @transient, are you sure you want them to become null after deserializing?
```
scala> case class X(@transient f: Option[Int] = None)
defined class X

scala> import java.util._
import java.util._

scala> import java.io._
import java.io._

scala> val bos = new ByteArrayOutputStream
bos: java.io.ByteArrayOutputStream =

scala> val os = new ObjectOutputStream(bos)
os: java.io.ObjectOutputStream = java.io.ObjectOutputStream@7f9fcf7f

scala> os.writeObject(new X(Some(123)))

scala> os.close()

scala> val bis = new ByteArrayInputStream(bos.toByteArray)
bis: java.io.ByteArrayInputStream = java.io.ByteArrayInputStream@22555ebf

scala> val is = new ObjectInputStream(bis)
is: java.io.ObjectInputStream = java.io.ObjectInputStream@131ef10

scala> val deserialized = is.readObject()
deserialized: Object = X(null)
//             ^^^^^^^^^^^^^^^^^^ see? null
```